stdlib/string.ml

```diff
@@ -280,3 +280,70 @@ let get_int32_le s i = B.get_int32_le (bos s) i
 let get_int32_be s i = B.get_int32_be (bos s) i
 let get_int64_le s i = B.get_int64_le (bos s) i
 let get_int64_be s i = B.get_int64_be (bos s) i
+
+(* Spellchecking *)
+
+let uchar_array_of_utf_8_string s =
+  let slen = length s in (* is an upper bound on Uchar.t count *)
+  let uchars = Array.make slen Uchar.max in
+  let k = ref 0 and i = ref 0 in
+  while (!i < slen) do
+    let dec = get_utf_8_uchar s !i in
+    i := !i + Uchar.utf_decode_length dec;
+    uchars.(!k) <- Uchar.utf_decode_uchar dec;
+    incr k;
+  done;
+  uchars, !k
+
+let edit_distance ?(limit = Int.max_int) s0 s1 =
+  let[@inline] minimum a b c = Int.min a (Int.min b c) in
+  let s0, len0 = uchar_array_of_utf_8_string s0 in
+  let s1, len1 = uchar_array_of_utf_8_string s1 in
+  let limit = Int.min (Int.max len0 len1) limit in
+  if Int.abs (len1 - len0) > limit then limit else
+  let s0, s1 = if len0 > len1 then s0, s1 else s1, s0 in
+  let len0, len1 = if len0 > len1 then len0, len1 else len1, len0 in
+  let rec loop row_minus2 row_minus1 row i len0 limit =
```

Comment:
Why are `len0` and `limit` parameters? Also, why is `len1` recomputed below?

Reply:
Because depending on how this gets compiled closure allocation may be avoided, `len1` is not recomputed, it is accessed. See [this discussion](https://discuss.ocaml.org/t/tail-recursive-loops-and-closures/10845).

Reply:
You are right. Simply because… I forgot! Though at that point we likely get to many vars for that manual paramter passing to be useful I should check goldbolt.